crates/winch/src/compiler.rs

```diff
@@ -1,50 +1,138 @@
 use anyhow::Result;
+use cranelift_codegen::{Final, MachBufferFinalized};
 use object::write::{Object, SymbolId};
 use std::any::Any;
+use std::sync::Mutex;
+use wasmparser::FuncValidatorAllocations;
+use wasmtime_cranelift_shared::obj::ModuleTextBuilder;
 use wasmtime_environ::{
-    CompileError, DefinedFuncIndex, FuncIndex, FunctionBodyData, FunctionLoc, ModuleTranslation,
-    ModuleTypes, PrimaryMap, Tunables, WasmFunctionInfo,
+    CompileError, DefinedFuncIndex, FilePos, FuncIndex, FunctionBodyData, FunctionLoc,
+    ModuleTranslation, ModuleTypes, PrimaryMap, Tunables, WasmFunctionInfo,
 };
 use winch_codegen::TargetIsa;
 
+use crate::environ::FuncEnv;
+
 pub(crate) struct Compiler {
     isa: Box<dyn TargetIsa>,
+    allocations: Mutex<Vec<FuncValidatorAllocations>>,
 }
 
+struct CompiledFunction(MachBufferFinalized<Final>);
+
 impl Compiler {
     pub fn new(isa: Box<dyn TargetIsa>) -> Self {
-        Self { isa }
+        Self {
+            isa,
+            allocations: Mutex::new(Vec::new()),
+        }
+    }
+
+    fn take_allocations(&self) -> FuncValidatorAllocations {
+        self.allocations
+            .lock()
+            .unwrap()
+            .pop()
+            .unwrap_or_else(Default::default)
+    }
+
+    fn save_allocations(&self, allocs: FuncValidatorAllocations) {
+        self.allocations.lock().unwrap().push(allocs)
     }
 }
 
 impl wasmtime_environ::Compiler for Compiler {
     fn compile_function(
         &self,
-        _translation: &ModuleTranslation<'_>,
-        _index: DefinedFuncIndex,
-        _data: FunctionBodyData<'_>,
+        translation: &ModuleTranslation<'_>,
+        index: DefinedFuncIndex,
+        data: FunctionBodyData<'_>,
         _tunables: &Tunables,
         _types: &ModuleTypes,
     ) -> Result<(WasmFunctionInfo, Box<dyn Any + Send>), CompileError> {
-        todo!()
+        let index = translation.module.func_index(index);
+        let sig = translation.get_types().function_at(index.as_u32()).unwrap();
+        let FunctionBodyData { body, validator } = data;
+        let start_srcloc = FilePos::new(
+            body.get_binary_reader()
+                .original_position()
+                .try_into()
+                .unwrap(),
+        );
+        let mut validator = validator.into_validator(self.take_allocations());
+        let env = FuncEnv::new(&translation.module, translation.get_types());
+        let buffer = self
+            .isa
+            .compile_function(&sig, &body, &env, &mut validator)
+            .map_err(|e| CompileError::Codegen(format!("{e:?}")));
+        self.save_allocations(validator.into_allocations());
+        let buffer = buffer?;
+
+        Ok((
+            WasmFunctionInfo {
+                start_srcloc,
+                stack_maps: Box::new([]),
+            },
+            Box::new(CompiledFunction(buffer)),
+        ))
     }
 
     fn compile_host_to_wasm_trampoline(
```

Comment:
Thanks for bringing this up @jameysharp. This is something that we've discussed during the Winch meetings. We are aware that these changes are coming with support for tail calls. 

> but we may call on you to help with the Winch side of that work once we've sorted out how the Cranelift version needs to go.

Absolutely. I'll note that since Winch is still in development we can easily adjust here moving forward.